### PR TITLE
Very simple deposit/withdraw logic and share accounting

### PR DESCRIPTION
**Pull request by mcclurejt**
- closes https://github.com/delvtech/hyperdrive/issues/1087
- closes https://github.com/delvtech/hyperdrive/issues/1088

Adds simple deposit + redeem logic.

Adds simple share accounting where each depositor receives shares equal to their deposit amount and share price is pegged to 1. This will be updated by future pr's.

### CODE REVIEW

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:46]: The best practice is generally to avoid dangling approvals. I get that we are going to do the rebalance anyways after deposit, but why don't we do the approval there before we use it? IMO that would make the code more readable.

Comment (jalextowle) [reply contracts/internal/EverlongERC4626.sol:46]: If we eliminate this, I think we can delete this function.

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:57]: Wouldn't it make more sense to close matured positions first? Unless I'm missing something, this is just going to pay out some idle to the user, which the pool won't have unless it closes existing positions. I understand why we'd want to put the remaining amount of capital into new positions afterwards, so we'd probably still keep this call afterwards.

Something else that I'd say here is that I get that the logic is simple, but each of these lines of code is pretty consequential and the order matters. IMO, both of these lines should be commented with a high-level description of what's happening and why it needs to happen here.

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:51]: A question that I'd have about this function is whether or not we're going to include withdrawal shares in this implementation. I think that it would really hurt the UX not to have them.

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:104]: This should be commented, and we should add a newline. Based on the style guide, there shouldn't be lines of code like this immediately above a comment block.

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:116]: My understanding of the initial spec was that we'd never close non-matured positions.

File: contracts/internal/EverlongERC4626.sol
Comment (jalextowle) [inline contracts/internal/EverlongERC4626.sol:16]: The way that we solved this kind of inheritance problem in Hyperdrive is to have the base contract contain an abstract function like `rebalance` (we did this with `checkpoint`). I prefer that style to the inheritance changes I'm seeing in this PR since the structure is flatter and easier to reason about.

File: contracts/internal/EverlongPositions.sol
Comment (jalextowle) [inline contracts/internal/EverlongPositions.sol:174]: As mentioned elsewhere, I thought we all agreed to not closing any positions before maturity. Doing so significantly increases the complexity of dealing with things like sandwich attacks.

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:49]: Should this be immutable? When would we change this?

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:57]: Should this be immutable? When would we change this?

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:63]: What happened to kind and version?

File: contracts/Everlong.sol
Comment (jalextowle) [inline contracts/Everlong.sol:63]: IMO this structure isn't as clear as the previous structure. Hyperdrive mostly follows a mixins pattern where each block of individual logic is implemented in a separate contract, and the feature contracts are combined to create the higher level interface. The nice thing about this pattern is that it's clear what the logic contracts are just from looking at the entrypoint. 

Another thought is that it's convenient to have abstract functions used by several mixins and internal functions used in several mixins in the base contract. Using this inheritance hierarchy makes that impossible and also makes the order of inheritance between the mixins more consequential.

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:25]: Shouldn't this be an immutable?

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:29]: Same here, when would this change?

File: test/units/Everlong.t.sol
Comment (jalextowle) [inline test/units/Everlong.t.sol:24]: Why don't we compare this to the constant `EVERLONG_VERSION`?

File: test/units/Everlong.t.sol
Comment (jalextowle) [inline test/units/Everlong.t.sol:19]: Why don't we compare this to the constant for kind?

File: test/units/EverlongERC4626.t.sol
Comment (jalextowle) [inline test/units/EverlongERC4626.t.sol:21]: Looks like there was a typo here:

```suggestion
    ///  @dev When calling the `deposit()` function...
    ///       1. The `Rebalanced` event should be emitted
    ///       2. The balance of the Everlong contract should be minimal.
    ///       3. The share amount issued to the depositor should be equal to the
    ///           total supply of Everlong shares.
    ///       4. The amount of assets deposited should relate to the share
    ///           count as follows:
    ///           shares_issued = assets_deposited * (10 ^ decimalsOffset)
```

File: test/units/EverlongERC4626.t.sol
Comment (jalextowle) [inline test/units/EverlongERC4626.t.sol:164]: We have this in another of our test files. Looks like this is duplicated.

File: test/units/EverlongERC4626.t.sol
Comment (jalextowle) [inline test/units/EverlongERC4626.t.sol:110]: Again, I remember talking about not doing this to reduce complexity/increase security.

**Review by jalextowle — COMMENTED** [review]
I left some comments. The main comments have to do with the decision to close immature positions. 

Could you also add a PR description? Those will be helpful when we need to give auditors or other parties documentation about Everlong and the decisions that were made.

File: contracts/internal/EverlongERC4626.sol
Comment (mcclurejt) [reply contracts/internal/EverlongERC4626.sol:46]: You're totally right, don't need this at all since there's already an approval in rebalance.

File: contracts/internal/EverlongERC4626.sol
Comment (mcclurejt) [reply contracts/internal/EverlongERC4626.sol:48]: I think that closing matured positions prior to redemption is handled in the [`_beforeWithdraw()`](https://github.com/delvtech/everlong/blob/4bac426c161cb0393b17fca902dcf064895c137e/contracts/internal/EverlongERC4626.sol#L98-L101) hook... but could be mistaken.

Will add high-level description + comments

File: test/units/EverlongERC4626.t.sol
Comment (mcclurejt) [reply test/units/EverlongERC4626.t.sol:21]: good eye!!!

File: contracts/internal/EverlongERC4626.sol
Comment (mcclurejt) [reply contracts/internal/EverlongERC4626.sol]: When I looked at Hyperdrive's `checkpoint` I didn't see any abstract functions used but maybe I'm missing it?

Regardless, I refactored to be more "flat" and is hopefully in line with what you meant. Currently running into the `Error (5005): Linearization of inheritance graph impossible` error I had mentioned before and pushed the repo in this state in case you want to take a look. 

The cause of this error is the addition of `IEverlong` to the list of extended items in `Everlong.sol`... seems very odd since none of the other contracts inherit `IEverlong`, they just reference errors from it.

Comment (mcclurejt) [reply contracts/internal/EverlongERC4626.sol]: Seems like the hyperdrive contracts never do `Hyperdrive is IHyperdrive` for this reason, so removing the interface from the inheritance.

File: contracts/internal/EverlongPositions.sol
Comment (mcclurejt) [reply contracts/internal/EverlongPositions.sol]: Resolving since our stance on this has changed.

File: contracts/Everlong.sol
Comment (mcclurejt) [reply contracts/Everlong.sol]: Have made changes in line with this to flatten things out and not have so much reliance on import ordering Let me know if there's more improvements you see! :)

File: contracts/Everlong.sol
Comment (jalextowle) [inline contracts/Everlong.sol:98]: Thinking about yield sources like `ChainlinkHyperdrive` or other kinds of bridged assets, it may be better to just have the deployer provide a `decimals` parameter.

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:28]: Why are these internal? We'll want getters for these functions.

File: contracts/internal/EverlongStorage.sol
Comment (jalextowle) [inline contracts/internal/EverlongStorage.sol:14]: As a general comment, I'm confused about why some things are public and others aren't. I understand why things like `asset` and `decimals` are internal since we need to override these getters due to the use of `ERC4626`, but there are several other fields that live entirely within Everlong that could be made public.

File: test/exposed/EverlongERC4626Exposed.sol
Comment (jalextowle) [inline test/exposed/EverlongERC4626Exposed.sol:11]: Nit:

```suggestion
    }
    
```

File: test/exposed/EverlongExposed.sol
Comment (jalextowle) [inline test/exposed/EverlongExposed.sol:6]: Nit: 

```suggestion
```

File: contracts/internal/EverlongPositions.sol
Comment (jalextowle) [inline contracts/internal/EverlongPositions.sol:177]: Not for this PR, but I think we'll want to re-evaluate our use of `inheritdoc`. If we're going to have documentation for this function, I'd want it to live here since this is where the function is implemented. Generally speaking, I prefer duplicated docs to not having any where the function is actually implemented.

File: contracts/internal/EverlongPositions.sol
Comment (jalextowle) [inline contracts/internal/EverlongPositions.sol:184]: We can update the style guide for this, but I prefer multi-line if-statements. The reason for this is that it makes it much easier to debug since you can add console statements _inside the if-statement_ to see what is happening.

```suggestion
        if (idle >= _target) {
            return idle;
        }
```

File: contracts/internal/EverlongPositions.sol
Comment (jalextowle) [inline contracts/internal/EverlongPositions.sol:189]: Same here.

```suggestion
        if (idle >= _target) {
            return idle;
        }
```

File: contracts/internal/EverlongPositions.sol
Comment (jalextowle) [inline contracts/internal/EverlongPositions.sol:210]:
```suggestion
            if (idle >= _target) {
                return idle;
            }
```

**Review by jalextowle — APPROVED** [review]
LGTM once the comments are addressed

File: contracts/internal/EverlongStorage.sol
Comment (mcclurejt) [reply contracts/internal/EverlongStorage.sol:14]: Good point, not super happy with how things are currently structured and was able to isolate this a bit better as part of the vault share pricing work. Will be cleaned up in next pr :)

File: contracts/Everlong.sol
Comment (mcclurejt) [reply contracts/Everlong.sol:98]: Totally agree, good callout

**Pull request merged**
